arangod/RocksDBEngine/RocksDBRecoveryManager.cpp

```diff
@@ -122,37 +128,41 @@ class WBReader final : public rocksdb::WriteBatch::Handler {
   explicit WBReader(std::unordered_map<uint64_t, rocksdb::SequenceNumber> const& seqs)
       : currentSeqNum(0), _seqStart(seqs) {}
 
-  ~WBReader() {
-    // update ticks after parsing wal
-    LOG_TOPIC(TRACE, Logger::ENGINES) << "max tick found in WAL: " << _maxTick
-                                      << ", last HLC value: " << _maxHLC;
-
-    TRI_UpdateTickServer(_maxTick);
-    TRI_HybridLogicalClock(_maxHLC);
-
-    // TODO update generators
-    auto dbfeature = ApplicationServer::getFeature<DatabaseFeature>("Database");
-    for (auto gen : _generators) {
-      if (gen.second > 0) {
-        auto dbColPair = rocksutils::mapObjectToCollection(gen.first);
-        if (dbColPair.second == 0 && dbColPair.first == 0) {
-          // collection with this objectID not known.Skip.
-          continue;
-        }
-        auto vocbase = dbfeature->useDatabase(dbColPair.first);
-        if (vocbase == nullptr) {
-          continue;
-        }
-        TRI_DEFER(vocbase->release());
+  Result shutdownWBReader() {
+    Result rv;
+    try {
+      // update ticks after parsing wal
+      LOG_TOPIC(TRACE, Logger::ENGINES) << "max tick found in WAL: " << _maxTick
+                                        << ", last HLC value: " << _maxHLC;
+
+      TRI_UpdateTickServer(_maxTick);
+      TRI_HybridLogicalClock(_maxHLC);
+
+      // TODO update generators
+      auto dbfeature = ApplicationServer::getFeature<DatabaseFeature>("Database");
+      for (auto gen : _generators) {
+        if (gen.second > 0) {
+          auto dbColPair = rocksutils::mapObjectToCollection(gen.first);
+          if (dbColPair.second == 0 && dbColPair.first == 0) {
+            // collection with this objectID not known.Skip.
+            continue;
+          }
+          auto vocbase = dbfeature->useDatabase(dbColPair.first);
+          if (vocbase == nullptr) {
+            continue;
+          }
+          TRI_DEFER(vocbase->release());
 
-        auto collection = vocbase->lookupCollection(dbColPair.second);
-        if (collection == nullptr) {
-          continue;
+          auto collection = vocbase->lookupCollection(dbColPair.second);
+          if (collection == nullptr) {
+            continue;
+          }
+          std::string k(basics::StringUtils::itoa(gen.second));
+          collection->keyGenerator()->track(k.data(), k.size());
         }
-        std::string k(basics::StringUtils::itoa(gen.second));
-        collection->keyGenerator()->track(k.data(), k.size());
       }
-    }
+    } CATCH_TO_RESULT(rv,TRI_ERROR_INTERNAL);
```

Comment:
Instead of using a macro, can we instead define a class whose constructor takes the result, a lambda, and default error (with a default value of `TRI_ERROR_INTERNAL`) and accomplishes the same? So then the idiom would be
```
TryCatchToResult(rv, [&] -> void {
  <content of try block>
});
```
Does the macro get us something that this method doesn't?

Reply:
Coffee hasn't fully kicked in yet. Could just be a function rather than a class/constructor.

Reply:
Shouldn't it rather be
`Result tryCatchToResult(std::function<Result()>);`, potentially with an overload for
`Result tryCatchToResult(std::function<void>());`?

That way we could use it in expressions, e.g.
`return tryCatchToResult([]() { ... return Result(); });` or `return tryCatchToResult([]() { bla(); });`.

Reply:
@jsteemann Yes, we could probably provide a few variants. In any case, we would want to handle a few of the known exception types specifically (e.g. `std::bad_alloc`).